sonar-batch/src/main/java/org/sonar/batch/issue/ModuleIssues.java

```diff
@@ -37,55 +46,126 @@
 
   private final ActiveRules activeRules;
   private final Rules rules;
-  private final IssueCache cache;
-  private final Project project;
   private final IssueFilters filters;
+  private final ReportPublisher reportPublisher;
+  private final BatchComponentCache componentCache;
+  private final BatchReport.Issue.Builder builder = BatchReport.Issue.newBuilder();
+  private final Builder locationBuilder = IssueLocation.newBuilder();
+  private final org.sonar.batch.protocol.output.BatchReport.TextRange.Builder textRangeBuilder = org.sonar.batch.protocol.output.BatchReport.TextRange.newBuilder();
+  private final BatchReport.ExecutionFlow.Builder flowBuilder = BatchReport.ExecutionFlow.newBuilder();
 
-  public ModuleIssues(ActiveRules activeRules, Rules rules, IssueCache cache, Project project, IssueFilters filters) {
+  public ModuleIssues(ActiveRules activeRules, Rules rules, IssueFilters filters, ReportPublisher reportPublisher, BatchComponentCache componentCache) {
     this.activeRules = activeRules;
     this.rules = rules;
-    this.cache = cache;
-    this.project = project;
     this.filters = filters;
+    this.reportPublisher = reportPublisher;
+    this.componentCache = componentCache;
   }
 
-  public boolean initAndAddIssue(DefaultIssue issue) {
-    RuleKey ruleKey = issue.ruleKey();
-    Rule rule = rules.find(ruleKey);
-    validateRule(issue, rule);
-    ActiveRule activeRule = activeRules.find(ruleKey);
+  public boolean initAndAddIssue(Issue issue) {
+    String key = ((DefaultInputComponent) issue.primaryLocation().inputComponent()).key();
+    BatchComponent component = componentCache.get(key);
+
+    Rule rule = validateRule(issue);
+    ActiveRule activeRule = activeRules.find(issue.ruleKey());
     if (activeRule == null) {
       // rule does not exist or is not enabled -> ignore the issue
       return false;
     }
-    updateIssue(issue, rule, activeRule);
-    if (filters.accept(issue)) {
-      cache.put(issue);
+
+    String primaryMessage = Strings.isNullOrEmpty(issue.primaryLocation().message()) ? rule.name() : issue.primaryLocation().message();
+    org.sonar.api.batch.rule.Severity overriddenSeverity = issue.overriddenSeverity();
+    Severity severity = overriddenSeverity != null ? Severity.valueOf(overriddenSeverity.name()) : Severity.valueOf(activeRule.severity());
+
+    builder.clear();
+    locationBuilder.clear();
+    // non-null fields
+    builder.setSeverity(severity);
+    builder.setRuleRepository(issue.ruleKey().repository());
+    builder.setRuleKey(issue.ruleKey().rule());
+    builder.setAttributes(KeyValueFormat.format(issue.attributes()));
+    builder.setMsg(primaryMessage);
+    locationBuilder.setMsg(primaryMessage);
+
+    locationBuilder.setComponentRef(component.batchId());
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) NullPointerException might be thrown as 'component' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2259)